// @alova/shared/dist/vars.cjs
/**
  * @alova/shared 1.0.5 (undefined)
  * Document undefined
  * Copyright 2024 Scott Hu. All Rights Reserved
  * Licensed under MIT (https://httpshub.com/alovajs/alova/blob/main/LICENSE)
*/

'use strict';

const undefStr = 'undefined';
// 以下为减少编译代码量而添加的统一处理函数或变量
const PromiseCls = Promise;
const promiseResolve = (value) => PromiseCls.resolve(value);
const promiseReject = (value) => PromiseCls.reject(value);
const ObjectCls = Object;
const RegExpCls = RegExp;
const undefinedValue = undefined;
const nullValue = null;
const trueValue = true;
const falseValue = false;
const promiseThen = (promise, onFulfilled, onrejected) => promise.then(onFulfilled, onrejected);
const promiseCatch = (promise, onrejected) => promise.catch(onrejected);
const promiseFinally = (promise, onfinally) => promise.finally(onfinally);
const JSONStringify = (value, replacer, space) => JSON.stringify(value, replacer, space);
const JSONParse = (value) => JSON.parse(value);
const setTimeoutFn = (fn, delay = 0) => setTimeout(fn, delay);
const clearTimeoutTimer = (timer) => clearTimeout(timer);
const objectKeys = (obj) => ObjectCls.keys(obj);
const objectValues = (obj) => ObjectCls.values(obj);
const forEach = (ary, fn) => ary.forEach(fn);
const pushItem = (ary, ...item) => ary.push(...item);
const mapItem = (ary, callbackfn) => ary.map(callbackfn);
const filterItem = (ary, predicate) => ary.filter(predicate);
const shift = (ary) => ary.shift();
const slice = (ary, start, end) => ary.slice(start, end);
const splice = (ary, start, deleteCount = 0, ...items) => ary.splice(start, deleteCount, ...items);
const len = (data) => data.length;
const isArray = (arg) => Array.isArray(arg);
const deleteAttr = (arg, attr) => delete arg[attr];
const typeOf = (arg) => typeof arg;
const regexpTest = (reg, str) => reg.test(str);
const includes = (ary, target) => ary.includes(target);
const valueObject = (value, writable = falseValue) => ({ value, writable });
const defineProperty = (o, key, value, isDescriptor = falseValue) => ObjectCls.defineProperty(o, key, isDescriptor ? value : valueObject(value, falseValue));
// 是否为服务端运行，node和bun通过process判断，deno通过Deno判断
// 部分框架（如支付宝和 uniapp）会注入 process 对象作为全局变量使用
// 因此使用服务端独有的 process.cwd 函数作为判断依据
const defaultIsSSR = typeof window === undefStr &&
    (typeof process !== undefStr ? typeof process.cwd === 'function' : typeof Deno !== undefStr);
const isSSR = defaultIsSSR;
/** cache mode */
// only cache in memory, it's default option
const MEMORY = 'memory';
// persistent cache, and will be read to memory when page is refreshed, it means that the memory cache always exist until cache is expired.
const STORAGE_RESTORE = 'restore';

exports.JSONParse = JSONParse;
exports.JSONStringify = JSONStringify;
exports.MEMORY = MEMORY;
exports.ObjectCls = ObjectCls;
exports.PromiseCls = PromiseCls;
exports.RegExpCls = RegExpCls;
exports.STORAGE_RESTORE = STORAGE_RESTORE;
exports.clearTimeoutTimer = clearTimeoutTimer;
exports.defaultIsSSR = defaultIsSSR;
exports.defineProperty = defineProperty;
exports.deleteAttr = deleteAttr;
exports.falseValue = falseValue;
exports.filterItem = filterItem;
exports.forEach = forEach;
exports.includes = includes;
exports.isArray = isArray;
exports.isSSR = isSSR;
exports.len = len;
exports.mapItem = mapItem;
exports.nullValue = nullValue;
exports.objectKeys = objectKeys;
exports.objectValues = objectValues;
exports.promiseCatch = promiseCatch;
exports.promiseFinally = promiseFinally;
exports.promiseReject = promiseReject;
exports.promiseResolve = promiseResolve;
exports.promiseThen = promiseThen;
exports.pushItem = pushItem;
exports.regexpTest = regexpTest;
exports.setTimeoutFn = setTimeoutFn;
exports.shift = shift;
exports.slice = slice;
exports.splice = splice;
exports.trueValue = trueValue;
exports.typeOf = typeOf;
exports.undefinedValue = undefinedValue;
exports.valueObject = valueObject;
