// alova/dist/stateshook/vue-demi.esm.js
/**
  * @alova/client 2.0.0 (https://alova.js.org)
  * Document https://alova.js.org
  * Copyright 2024 Scott hu. All Rights Reserved
  * Licensed under MIT (https://github.com/alovajs/alova/blob/main/LICENSE)
*/

import { setTimeoutFn, forEach, trueValue, clearTimeoutTimer, undefinedValue } from '@alova/shared/vars';
import { ref, getCurrentInstance, onUnmounted, onMounted, watch, computed } from 'vue-demi';

// Vue的预定义hooks
var vueDemi = {
    name: 'VueDemi',
    create: data => ref(data),
    dehydrate: state => state.value,
    update: (newVal, state) => {
        state.value = newVal;
    },
    effectRequest({ handler, removeStates, immediate, watchingStates }) {
        // 当在组件内部使用时，组件卸载时移除对应状态
        if (getCurrentInstance()) {
            onUnmounted(removeStates);
            onMounted(() => immediate && handler());
        }
        else {
            // 在非组件内部使用时，使用定时器延迟执行
            setTimeoutFn(() => {
                immediate && handler();
            });
        }
        let timer;
        forEach(watchingStates || [], (state, i) => {
            watch(state, () => {
                timer && clearTimeoutTimer(timer);
                timer = setTimeoutFn(() => {
                    handler(i);
                    timer = undefinedValue;
                });
            }, { deep: trueValue });
        });
    },
    computed: getter => computed(getter),
    watch: (states, callback) => {
        watch(states, callback, {
            deep: trueValue
        });
    },
    onMounted: callback => {
        onMounted(callback);
    },
    onUnmounted: callback => {
        onUnmounted(callback);
    }
};

export { vueDemi as default };
