// alova/dist/stateshook/vue-demi.umd.js
/**
  * @alova/client 2.0.0 (https://alova.js.org)
  * Document https://alova.js.org
  * Copyright 2024 Scott hu. All Rights Reserved
  * Licensed under MIT (https://github.com/alovajs/alova/blob/main/LICENSE)
*/

(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('vue-demi')) :
  typeof define === 'function' && define.amd ? define(['vue-demi'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, global.vueHook = factory(global.Vue));
})(this, (function (vueDemi$1) { 'use strict';

  /**
    * @alova/shared 1.0.5 (undefined)
    * Document undefined
    * Copyright 2024 Scott Hu. All Rights Reserved
    * Licensed under MIT (https://httpshub.com/alovajs/alova/blob/main/LICENSE)
  */

  const undefStr = 'undefined';
  const undefinedValue = undefined;
  const trueValue = true;
  const setTimeoutFn = (fn, delay = 0) => setTimeout(fn, delay);
  const clearTimeoutTimer = (timer) => clearTimeout(timer);
  const forEach = (ary, fn) => ary.forEach(fn);
  // 是否为服务端运行，node和bun通过process判断，deno通过Deno判断
  // 部分框架（如支付宝和 uniapp）会注入 process 对象作为全局变量使用
  // 因此使用服务端独有的 process.cwd 函数作为判断依据
  typeof window === undefStr &&
      (typeof process !== undefStr ? typeof process.cwd === 'function' : typeof Deno !== undefStr);

  // Vue的预定义hooks
  var vueDemi = {
      name: 'VueDemi',
      create: data => vueDemi$1.ref(data),
      dehydrate: state => state.value,
      update: (newVal, state) => {
          state.value = newVal;
      },
      effectRequest({ handler, removeStates, immediate, watchingStates }) {
          // 当在组件内部使用时，组件卸载时移除对应状态
          if (vueDemi$1.getCurrentInstance()) {
              vueDemi$1.onUnmounted(removeStates);
              vueDemi$1.onMounted(() => immediate && handler());
          }
          else {
              // 在非组件内部使用时，使用定时器延迟执行
              setTimeoutFn(() => {
                  immediate && handler();
              });
          }
          let timer;
          forEach(watchingStates || [], (state, i) => {
              vueDemi$1.watch(state, () => {
                  timer && clearTimeoutTimer(timer);
                  timer = setTimeoutFn(() => {
                      handler(i);
                      timer = undefinedValue;
                  });
              }, { deep: trueValue });
          });
      },
      computed: getter => vueDemi$1.computed(getter),
      watch: (states, callback) => {
          vueDemi$1.watch(states, callback, {
              deep: trueValue
          });
      },
      onMounted: callback => {
          vueDemi$1.onMounted(callback);
      },
      onUnmounted: callback => {
          vueDemi$1.onUnmounted(callback);
      }
  };

  return vueDemi;

}));
