// alova/dist/stateshook/vue.common.cjs
/**
  * @alova/client 2.0.0 (https://alova.js.org)
  * Document https://alova.js.org
  * Copyright 2024 Scott hu. All Rights Reserved
  * Licensed under MIT (https://github.com/alovajs/alova/blob/main/LICENSE)
*/

'use strict';

var _function = require('@alova/shared/function');
var vars = require('@alova/shared/vars');
var vue$1 = require('vue');

// Vue的预定义hooks
var vue = {
    name: 'Vue',
    create: data => vue$1.ref(data),
    dehydrate: state => state.value,
    update: (newVal, state) => {
        state.value = newVal;
    },
    effectRequest({ handler, removeStates, immediate, watchingStates }) {
        // 当在组件内部使用时，组件卸载时移除对应状态
        if (vue$1.getCurrentInstance()) {
            vue$1.onUnmounted(removeStates);
            vue$1.onMounted(() => immediate && handler());
        }
        else {
            // 在非组件内部使用时，使用定时器延迟执行
            vars.setTimeoutFn(() => {
                immediate && handler();
            });
        }
        const syncRunner = _function.createSyncOnceRunner();
        vars.forEach(watchingStates || [], (state, i) => {
            vue$1.watch(state, () => {
                syncRunner(() => {
                    handler(i);
                });
            }, { deep: vars.trueValue });
        });
    },
    computed: getter => vue$1.computed(getter),
    watch: (states, callback) => {
        vue$1.watch(states, callback, {
            deep: vars.trueValue
        });
    },
    onMounted: callback => {
        if (vue$1.getCurrentInstance()) {
            vue$1.onMounted(callback);
        }
        else {
            vars.setTimeoutFn(callback, 10);
        }
    },
    onUnmounted: callback => {
        vue$1.getCurrentInstance() && vue$1.onUnmounted(callback);
    }
};

module.exports = vue;
